website/ref/language.md

```diff
@@ -55,30 +55,70 @@ some values. (The traditional terms for the two levels are "commands" and
 ## String
 
 The most common data structure in shells is the string. String literals can be
-quoted or unquoted (barewords).
+quoted or unquoted (barewords). There are two types of quoted strings in Elvish:
+single-quoted strings and double-quoted strings.
 
-### Quoted
+### Single-Quoted String
 
-There are two types of quoted strings in Elvish, single-quoted strings and
-double-quoted strings.
-
-In single-quoted strings, all characters represent themselves, except single
+In single-quoted strings all characters represent themselves, except single
 quotes, which need to be doubled. For instance, `'*\'` evaluates to `*\`, and
 `'it''s'` evaluates to `it's`.
 
-In double-quoted strings, the backslash `\` introduces a **escape sequence**.
-For instance, `"\n"` evaluates to a newline; `"\\"` evaluates to a backslash;
-invalid escape sequences like `"\*"` result in a syntax error.
+### Double-Quoted String
 
-**TODO**: Document the full list of supported escape sequences.
+In double-quoted strings the backslash, `\`, introduces an **escape sequence**.
+For instance, `\n` evaluates to a newline and `\\` evaluates to a backslash.
+Invalid escape sequences like `\*` result in a syntax error when the program is
+compiled.
 
-Unlike most other shells, double-quoted strings do not support interpolation.
-For instance, `"$USER"` simply evaluates to the string `$USER`. To get a similar
-effect, simply concatenate strings: instead of `"my name is $name"`, write
-`"my name is "$name`. Under the hood this is a
+Unlike most other shells, double-quoted strings in elvish do not support
+interpolation. For instance, `"$name"` simply evaluates to the string `$name`.
+To get a similar effect, simply concatenate strings: instead of
+`"my name is $name"`, write `"my name is "$name`. Under the hood this is a
 [compound expression](#compound-expression-and-braced-lists).
 
-### Barewords
+These are the recognized escape sequences in double-quoted strings:
+
+-   `\cX` converts `X` to an ASCII control character. For example, `\cI` is the
```

Comment:
Defining this in terms of control characters introduces an additional concept unnecessarily, and I'd prefer this just to be defined in terms of the mechanism (substract 0x40 from the codepoint), and say that this corresponds to control characters as a non-normative comment:

> \cX, where `X` is a character with codepoint between 0x40 and 0x5F, represents the codepoint that is 0x40 lower than X. This corresponds to control character annotations in Unix.

Reply:
Your wording is arguably better but less flexible. That the implementation defines it in terms of subtracting 0x40 is not consistent with the historical definition. Specifically, masking the character with 0x1F so that `\ca` and `\cA` are equivalent. I was surprised when looking at the implementation to find that `\ca` isn't legal. My original text did mention masking with 0x1F. I probably should have retained that text. I would argue that the implementation should be relaxed to allow lowercase letters.

Reply:
Note that bash, as just one example, uses `#define TOCTRL(x) ((x) == '?' ? 0x7f : (TOUPPER(x) & 0x1f))` and `#define control_character_mask 0x1f`. Historically, various terminals disallowed applying the control modifier to some characters (such as `;`) and, as bash does, treat `\c?` specially as a synonym for 0x7F (del) rather than just masking `?` with 0x1F to produce 0x0F (si).

Reply:
Unsurprisingly, Wikipedia has a pretty decent [article](https://en.wikipedia.org/wiki/Control_character) concerning ASCII control characters. My opinion is that Elvish should handle this similar to bash but with `\c ` (i.e., control space) as a special-case that maps to `\c@` (aka nul). Just as `\c?` is a special-case for 0x7F (aka del). That should provide maximal compatibility with existing implementations and least surprise for users. Especially with respect to upper versus lowercase letters.